Bottom: 98d2d6e0ab85fb3832f7df44d89642c25137ad15
Top:    726c60bb5bdc68581a2f4a63004ab2cab4bda8b9
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-03-19 20:40:27 +0800

Refresh of drop-vhost_copy_from

---

diff --git a/drivers/vhost/vhost.c b/drivers/vhost/vhost.c
index a80346a..9ad9c1e 100644
--- a/drivers/vhost/vhost.c
+++ b/drivers/vhost/vhost.c
@@ -2000,7 +2000,7 @@ int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 {
 	int ret = 0;
 	u16 last_avail_idx, total;
-	__virtio16 avail_idx;
+	__virtio16 avail_idx, head;
 
 	if (unlikely(vhost_get_avail(vq, avail_idx, &vq->avail->idx))) {
 		vq_err(vq, "Failed to access avail idx at %p\n",
@@ -2010,16 +2010,17 @@ int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 	last_avail_idx = vq->last_avail_idx;
 	vq->avail_idx = vhost16_to_cpu(vq, avail_idx);
 	total = vq->avail_idx - vq->last_avail_idx;
+	total = min(total, num);
 
 	while (total) {
-		int ret2 = vhost_get_avail(vq, indices[ret],
+		int ret2 = vhost_get_avail(vq, head,
 			   &vq->avail->ring[last_avail_idx & (vq->num - 1)]);
 		if (unlikely(ret2)) {
 			vq_err(vq, "Failed to get descriptors\n");
 			return -EFAULT;
 		}
+		indices[ret++] = head;
 		--total;
-		++ret;
 		++last_avail_idx;
 	}
 	/* FIXME: update used ring here? together with batch dequing? */
